File: api/indicators.py
"""Technical indicators computed over native OHLCV candles.

Pure Python, no third-party deps. Series are aligned to the candle list;
warm-up positions are None. Donchian bands exclude the current bar (prior-N
window) so "close crosses above the upper band" is actually detectable —
a close can never exceed the high of its own bar.
"""

from __future__ import annotations

from typing import TypedDict


class Candle(TypedDict):
    time: int  # unix seconds, bar open
    open: float
    high: float
    low: float
    close: float
    volume: float


def ema(values: list[float], period: int) -> list[float | None]:
    out: list[float | None] = [None] * len(values)
    if len(values) < period:
        return out
    k = 2 / (period + 1)
    seed = sum(values[:period]) / period
    out[period - 1] = seed
    prev = seed
    for i in range(period, len(values)):
        prev = values[i] * k + prev * (1 - k)
        out[i] = prev
    return out


def sma(values: list[float], period: int) -> list[float | None]:
    out: list[float | None] = [None] * len(values)
    if len(values) < period:
        return out
    total = sum(values[:period])
    out[period - 1] = total / period
    for i in range(period, len(values)):
        total += values[i] - values[i - period]
        out[i] = total / period
    return out


def rsi(closes: list[float], period: int = 14) -> list[float | None]:
    out: list[float | None] = [None] * len(closes)
    if len(closes) <= period:
        return out
    gains = losses = 0.0
    for i in range(1, period + 1):
        delta = closes[i] - closes[i - 1]
        gains += max(delta, 0.0)
        losses += max(-delta, 0.0)
    avg_gain = gains / period
    avg_loss = losses / period

    def value(g: float, loss: float) -> float:
        if loss == 0:
            return 100.0
        rs = g / loss
        return 100.0 - 100.0 / (1.0 + rs)

    out[period] = value(avg_gain, avg_loss)
    for i in range(period + 1, len(closes)):
        delta = closes[i] - closes[i - 1]
        avg_gain = (avg_gain * (period - 1) + max(delta, 0.0)) / period
        avg_loss = (avg_loss * (period - 1) + max(-delta, 0.0)) / period
        out[i] = value(avg_gain, avg_loss)
    return out


def atr(candles: list[Candle], period: int = 14) -> list[float | None]:
    out: list[float | None] = [None] * len(candles)
    if len(candles) <= period:
        return out
    trs: list[float] = []
    for i in range(1, len(candles)):
        c, prev_close = candles[i], candles[i - 1]["close"]
        trs.append(
            max(c["high"] - c["low"], abs(c["high"] - prev_close), abs(c["low"] - prev_close))
        )
    prev = sum(trs[:period]) / period
    out[period] = prev
    for i in range(period + 1, len(candles)):
        prev = (prev * (period - 1) + trs[i - 1]) / period
        out[i] = prev
    return out


def donchian(candles: list[Candle], period: int = 20) -> dict[str, list[float | None]]:
    """Prior-N-bar channel (excludes the current bar)."""
    n = len(candles)
    upper: list[float | None] = [None] * n
    lower: list[float | None] = [None] * n
    mid: list[float | None] = [None] * n
    for i in range(period, n):
        window = candles[i - period : i]
        hi = max(c["high"] for c in window)
        lo = min(c["low"] for c in window)
        upper[i], lower[i], mid[i] = hi, lo, (hi + lo) / 2
    return {"upper": upper, "lower": lower, "mid": mid}


def volume_ratio(candles: list[Candle], period: int = 20) -> float | None:
    """Latest bar volume vs the average of the prior `period` bars."""
    if len(candles) < period + 1:
        return None
    prior = candles[-period - 1 : -1]
    avg = sum(c["volume"] for c in prior) / period
    if avg == 0:
        return None
    return candles[-1]["volume"] / avg


def support_resistance(
    candles: list[Candle], lookback: int = 100, tolerance_pct: float = 0.5
) -> dict[str, list[dict]]:
    """Swing-point clustering: local highs/lows (2-bar wings) over the last
    `lookback` bars, clustered within `tolerance_pct` of price."""
    window = candles[-lookback:]
    swings: list[tuple[float, str]] = []
    for i in range(2, len(window) - 2):
        highs = [window[j]["high"] for j in (i - 2, i - 1, i + 1, i + 2)]
        lows = [window[j]["low"] for j in (i - 2, i - 1, i + 1, i + 2)]
        if window[i]["high"] > max(highs):
            swings.append((window[i]["high"], "resistance"))
        if window[i]["low"] < min(lows):
            swings.append((window[i]["low"], "support"))

    clusters: list[dict] = []
    for price, kind in sorted(swings):
        placed = False
        for cluster in clusters:
            ref = cluster["level"]
            if ref and abs(price - ref) / ref * 100 <= tolerance_pct:
                cluster["prices"].append(price)
                cluster["level"] = sum(cluster["prices"]) / len(cluster["prices"])
                cluster["touches"] += 1
                placed = True
                break
        if not placed:
            clusters.append({"level": price, "prices": [price], "touches": 1, "kind": kind})

    last_close = candles[-1]["close"]
    supports = sorted(
        (
            {"level": round(c["level"], 6), "touches": c["touches"]}
            for c in clusters
            if c["level"] < last_close
        ),
        key=lambda c: -c["level"],
    )[:4]
    resistances = sorted(
        (
            {"level": round(c["level"], 6), "touches": c["touches"]}
            for c in clusters
            if c["level"] > last_close
        ),
        key=lambda c: c["level"],
    )[:4]
    return {"supports": supports, "resistances": resistances}

File: site/lib/indicators.ts
import type { Candle, Indicators, Level } from "./types";

/**
 * Deterministic technical indicators over native OHLCV candles.
 *
 * Native candles only. No Heikin-Ashi, no smoothing of the price series before
 * it reaches these functions — that is the first of the four death traps and it
 * is rejected at the data layer, not here.
 */

export function sma(values: number[], period: number): (number | null)[] {
  const out: (number | null)[] = new Array(values.length).fill(null);
  if (values.length < period) return out;
  let total = 0;
  for (let i = 0; i < period; i++) total += values[i];
  out[period - 1] = total / period;
  for (let i = period; i < values.length; i++) {
    total += values[i] - values[i - period];
    out[i] = total / period;
  }
  return out;
}

/** Wilder's RSI. */
export function rsi(closes: number[], period = 14): (number | null)[] {
  const out: (number | null)[] = new Array(closes.length).fill(null);
  if (closes.length <= period) return out;

  let gains = 0;
  let losses = 0;
  for (let i = 1; i <= period; i++) {
    const d = closes[i] - closes[i - 1];
    if (d > 0) gains += d;
    else losses -= d;
  }
  let avgGain = gains / period;
  let avgLoss = losses / period;

  const value = (g: number, l: number) => (l === 0 ? 100 : 100 - 100 / (1 + g / l));
  out[period] = value(avgGain, avgLoss);

  for (let i = period + 1; i < closes.length; i++) {
    const d = closes[i] - closes[i - 1];
    avgGain = (avgGain * (period - 1) + Math.max(d, 0)) / period;
    avgLoss = (avgLoss * (period - 1) + Math.max(-d, 0)) / period;
    out[i] = value(avgGain, avgLoss);
  }
  return out;
}

/** Wilder's ATR over true range. */
export function atr(candles: Candle[], period = 14): (number | null)[] {
  const out: (number | null)[] = new Array(candles.length).fill(null);
  if (candles.length <= period) return out;

  const tr: number[] = [];
  for (let i = 1; i < candles.length; i++) {
    const c = candles[i];
    const prevClose = candles[i - 1].close;
    tr.push(
      Math.max(
        c.high - c.low,
        Math.abs(c.high - prevClose),
        Math.abs(c.low - prevClose)
      )
    );
  }

  let prev = tr.slice(0, period).reduce((a, b) => a + b, 0) / period;
  out[period] = prev;
  for (let i = period + 1; i < candles.length; i++) {
    prev = (prev * (period - 1) + tr[i - 1]) / period;
    out[i] = prev;
  }
  return out;
}

/**
 * Prior-N-bar Donchian channel, excluding the current bar — so "close crosses
 * above the upper band" is actually detectable. A close can never exceed the
 * high of its own bar.
 */
export function donchianUpper(candles: Candle[], period = 20): (number | null)[] {
  const out: (number | null)[] = new Array(candles.length).fill(null);
  for (let i = period; i < candles.length; i++) {
    let hi = -Infinity;
    for (let j = i - period; j < i; j++) hi = Math.max(hi, candles[j].high);
    out[i] = hi;
  }
  return out;
}

/** Swing-point clustering: local extremes with 2-bar wings, merged by proximity. */
export function supportResistance(
  candles: Candle[],
  lookback = 140,
  tolerancePct = 0.6
): { supports: Level[]; resistances: Level[] } {
  const w = candles.slice(-lookback);
  const swings: number[] = [];

  for (let i = 2; i < w.length - 2; i++) {
    const highs = [w[i - 2].high, w[i - 1].high, w[i + 1].high, w[i + 2].high];
    const lows = [w[i - 2].low, w[i - 1].low, w[i + 1].low, w[i + 2].low];
    if (w[i].high > Math.max(...highs)) swings.push(w[i].high);
    if (w[i].low < Math.min(...lows)) swings.push(w[i].low);
  }

  const clusters: { level: number; prices: number[] }[] = [];
  for (const price of swings.sort((a, b) => a - b)) {
    const hit = clusters.find(
      (c) => Math.abs(price - c.level) / c.level * 100 <= tolerancePct
    );
    if (hit) {
      hit.prices.push(price);
      hit.level = hit.prices.reduce((a, b) => a + b, 0) / hit.prices.length;
    } else {
      clusters.push({ level: price, prices: [price] });
    }
  }

  const last = candles[candles.length - 1].close;
  const toLevel = (c: { level: number; prices: number[] }): Level => ({
    price: round(c.level),
    touches: c.prices.length,
  });

  return {
    supports: clusters
      .filter((c) => c.level < last)
      .sort((a, b) => b.level - a.level)
      .slice(0, 4)
      .map(toLevel),
    resistances: clusters
      .filter((c) => c.level > last)
      .sort((a, b) => a.level - b.level)
      .slice(0, 4)
      .map(toLevel),
  };
}

function round(n: number): number {
  const abs = Math.abs(n);
  const dp = abs >= 1000 ? 2 : abs >= 10 ? 3 : abs >= 1 ? 4 : 6;
  return Number(n.toFixed(dp));
}

export function computeIndicators(candles: Candle[]): Indicators {
  const closes = candles.map((c) => c.close);
  const ma20 = sma(closes, 20);
  const ma50 = sma(closes, 50);
  const rsiSeries = rsi(closes, 14);
  const atrSeries = atr(candles, 14);

  const last = closes[closes.length - 1];
  const prev = closes[closes.length - 2] ?? last;
  const m20 = ma20[ma20.length - 1];
  const m50 = ma50[ma50.length - 1];
  const atrLast = atrSeries[atrSeries.length - 1];

  const maSpreadPct = m20 != null && m50 != null ? ((m20 - m50) / m50) * 100 : null;
  let trend: Indicators["trend"] = "ranging";
  if (maSpreadPct != null) {
    if (maSpreadPct > 0.35) trend = "uptrend";
    else if (maSpreadPct < -0.35) trend = "downtrend";
  }

  // Latest bar's volume against the mean of the prior 20 bars.
  let volumeRatio: number | null = null;
  if (candles.length > 21) {
    const prior = candles.slice(-21, -1);
    const avg = prior.reduce((a, c) => a + c.volume, 0) / prior.length;
    volumeRatio = avg > 0 ? candles[candles.length - 1].volume / avg : null;
  }

  const { supports, resistances } = supportResistance(candles);

  return {
    last: round(last),
    changePct: Number((((last - prev) / prev) * 100).toFixed(2)),
    ma20,
    ma50,
    rsi: rsiSeries,
    rsiLast: rsiSeries[rsiSeries.length - 1],
    atrLast: atrLast != null ? round(atrLast) : null,
    atrPct: atrLast != null ? Number(((atrLast / last) * 100).toFixed(2)) : null,
    volumeRatio: volumeRatio != null ? Number(volumeRatio.toFixed(2)) : null,
    trend,
    maSpreadPct: maSpreadPct != null ? Number(maSpreadPct.toFixed(2)) : null,
    supports,
    resistances,
  };
}
